src/tests/resource-listers/AllResourcesLister.test.js

```diff
@@ -0,0 +1,210 @@
+'use strict';
+
+var _ = require('underscore'),
+    Q = require('q'),
+    sinon = require('sinon'),
+    expect = require('expect.js'),
+    rewire = require('rewire'),
+    Lister = rewire('../../resource-listers/AllResourcesLister');
+
+describe('AllResourcesLister', function() {
+   var dynamo = { listTables: _.noop },
+       lister, revert;
+
+   beforeEach(function() {
+      revert = Lister.__set__({
+         dynamo: dynamo,
+         console: { log: _.noop },
+      });
+
+      lister = new Lister();
+   });
+
+   afterEach(function() {
+      revert();
+   });
+
+   describe('_fetchTableNames', function() {
+
+      it('correctly fetches a list of tables - integration', function() {
+         var stub = sinon.stub(dynamo, 'listTables'),
+             expectedResponse1 = { TableNames: [ 'table-a', 'table-b', 'table-c' ], LastEvaluatedTableName: 'next-table-set' },
+             expectedResponse2 = { TableNames: [ 'table-d', 'table-e', 'table-f' ], LastEvaluatedTableName: 'further-table-set' },
+             expectedResponse3 = { TableNames: [ 'table-g', 'table-h', 'table-i' ] };
+
+         stub.onCall(0).callsArgWithAsync(1, null, expectedResponse1);
+         stub.onCall(1).callsArgWithAsync(1, null, expectedResponse2);
+         stub.onCall(2).callsArgWithAsync(1, null, expectedResponse3);
+
+         return lister._fetchTableNames()
+            .then(function(tableNames) {
+               sinon.assert.calledThrice(stub);
+               sinon.assert.calledOn(stub, dynamo);
+
+               sinon.assert.calledWith(stub, { ExclusiveStartTableName: null });
+               sinon.assert.calledWith(stub, { ExclusiveStartTableName: 'next-table-set' });
+               sinon.assert.calledWith(stub, { ExclusiveStartTableName: 'further-table-set' });
+               sinon.assert.callOrder(
+                  stub.withArgs({ ExclusiveStartTableName: null }),
+                  stub.withArgs({ ExclusiveStartTableName: 'next-table-set' }),
+                  stub.withArgs({ ExclusiveStartTableName: 'further-table-set' })
+               );
+
+               expect(tableNames).to.eql([
+                  'table-a', 'table-b', 'table-c',
+                  'table-d', 'table-e', 'table-f',
+                  'table-g', 'table-h', 'table-i',
+               ]);
+
+               stub.restore();
+            });
+      });
+
+      it('processes data from _fetchAll properly', function() {
+         var stub = sinon.stub(lister, '_fetchAll');
+
+         stub.returns(Q.when([
```

Comment:
Instead of `Q.when`, use `Q.delay` with a few milliseconds of delay ... I typically just use three milliseconds. I find that doing this helps make the tests more correct and realistic, and has exposed bugs before. Using `Q.when` doesn't truly test the async nature of the code since it resolves immediately and therefore does not have the same effect on the event loop as an async op would.